creme/tree/_base_tree.py

```diff
@@ -0,0 +1,411 @@
+        attribute at any given node.
+        $n$: Number of samples.
+
+        Parameters
+        ----------
+        range_val
+            Range value.
+        confidence
+            Confidence of choosing the correct attribute.
+        n
+            Number of processed samples.
+        """
+        return math.sqrt((range_val * range_val * math.log(1. / confidence)) / (2. * n))
+
+    @property
+    def max_size(self):
+        return self._max_size
+
+    @max_size.setter
+    def max_size(self, size):
+        self._max_size = size
+        self._max_byte_size = self._max_size * (2 ** 20)
+
+    @property
+    def model_measurements(self):
+        """Collect metrics corresponding to the current status of the tree.
+
+        Returns
+        -------
+        string
+            A string buffer containing the measurements of the tree.
+        """
+        measurements = {'Tree size (nodes)': self._n_decision_nodes
+                        + self._n_active_leaves + self._n_inactive_leaves,
+                        'Tree size (leaves)': self._n_active_leaves
+                        + self._n_inactive_leaves,
+                        'Active learning nodes': self._n_active_leaves,
+                        'Tree depth': self.depth,
+                        'Active leaf byte size estimate': self._active_leaf_size_estimate,
+                        'Inactive leaf byte size estimate': self._inactive_leaf_size_estimate,
+                        'Byte size estimate overhead': self._size_estimate_overhead_fraction
+                        }
+        return measurements
+
+    def get_model_description(self):
+        """Walk the tree and return its structure in a buffer.
+
+        Returns
+        -------
+        string
+            The description of the model.
+
+        """
+        if self._tree_root is not None:
+            buffer = ['']
+            description = ''
+            self._tree_root.describe_subtree(self, buffer, 0)
+            for line in range(len(buffer)):
+                description += buffer[line]
+            return description
+
+    def _new_split_node(self, split_test: InstanceConditionalTest, target_stats: dict = None,
+                        depth: int = 0) -> SplitNode:
+        """Create a new split node."""
+        return SplitNode(split_test, target_stats, depth)
+
+    @abstractmethod
+    def _new_learning_node(self, initial_stats: dict = None, parent: LearningNode = None,
+                           is_active: bool = True) -> LearningNode:
+        """Create a new learning node.
+
+        The characteristics of tje learning node depends on the tree algorithm.
+
+        Parameters
+        ----------
+        initial_stats
+            Target statistics inherited from the parent node.
+        parent
+            Parent node to inherit from.
+        is_active
+            Define whether or not the new node to be created is an active learning node.
+
+        Returns
+        -------
+        node
```

Comment:
That's a good point. I am not sure whether or not I should use the types there. I ended up removing `node` because the method's signature already has `(...) -> LearningNode:`.

Perhaps I'll need to give a complete scan in the module to remove all the occurences of return types in the docstring. 

Any thoughts, @MaxHalford?

Reply:
Yeah the ideal would be to indicate the return annotation type instead of specifying the type as a comment. The comment should just indicate additional information.